mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/SmartTestingMavenConfigurer.java

```diff
@@ -133,16 +139,56 @@ private void calculateChanges(File projectDirectory, Configuration configuration
     }
 
     private void configureExtension(MavenSession session, Configuration configuration) {
-        final MavenProjectConfigurator mavenProjectConfigurator = new MavenProjectConfigurator(configuration);
-        session.getAllProjects().forEach(mavenProject -> {
-            boolean wasConfigured = mavenProjectConfigurator.configureTestRunner(mavenProject.getModel());
-            if (wasConfigured) {
-                configuration.dump(mavenProject.getBasedir());
-                if (isFailedStrategyUsed()) {
-                    SurefireReportStorage.copySurefireReports(mavenProject.getModel());
-                }
+        final Consumer<MavenProject> configureSmartTestingExtensionAction;
+        final File executionRootDirectory = new File(session.getExecutionRootDirectory());
+        if (hasModuleSpecificConfigurations(executionRootDirectory, this::isProjectRootDirectory)) {
+            configureSmartTestingExtensionAction = applyModuleSpecificConfiguration(configuration);
+        } else {
+            final MavenProjectConfigurator mavenProjectConfigurator = new MavenProjectConfigurator(configuration);
+            configureSmartTestingExtensionAction = mavenProject -> configureMavenProject(mavenProjectConfigurator, mavenProject, configuration);
+        }
+        session.getAllProjects().forEach(configureSmartTestingExtensionAction);
+    }
+
+    private boolean hasModuleSpecificConfigurations(File projectDir, Function<File, Boolean> stopCondition) {
+        final ConfigLookup lookUp = new ConfigLookup(projectDir, stopCondition);
+        final boolean moreThanOneConfigFile = lookUp.hasMoreThanOneConfigFile(SMART_TESTING_YML, SMART_TESTING_YAML);
+        return moreThanOneConfigFile && System.getProperty(SMART_TESTING_CONFIG) == null;
+    }
+
+    private Consumer<MavenProject> applyModuleSpecificConfiguration(Configuration configuration) {
+        return mavenProject -> {
+            final ConfigLookup configLookup = new ConfigLookup(mavenProject.getBasedir(), this::isProjectRootDirectory);
+            final File dirWithConfig = configLookup.getFirstDirWithConfigOrWithStopCondition();
+            final Configuration mavenProjectConfiguration = configLookup.isConfigFromProjectRootDir() ? configuration :
+                ConfigurationLoader.load(dirWithConfig);
+            if (mavenProjectConfiguration.isDisable()) {
+                logExtensionDisableReason(logger,
+                    SMART_TESTING_DISABLE + " is set for module " + mavenProject.getArtifactId());
+                return;
             }
-        });
+            if (mavenProjectConfiguration.areStrategiesDefined()) {
+                final MavenProjectConfigurator mavenProjectConfigurator =
+                    new MavenProjectConfigurator(mavenProjectConfiguration);
```

Comment:
the instantiation of this class can be inside of the `configureMavenProject` method - you pass the configuration instance there anyway